src/SingleReport/MultiTouch.cpp

```diff
@@ -0,0 +1,246 @@
+	};
+	return USB_SendControl(0, &hidInterface, sizeof(hidInterface));
+}
+
+int MultiTouch_::getDescriptor(USBSetup& setup)
+{
+	// Check if this is a HID Class Descriptor request
+	if (setup.bmRequestType != REQUEST_DEVICETOHOST_STANDARD_INTERFACE) { return 0; }
+	if (setup.wValueH != HID_REPORT_DESCRIPTOR_TYPE) { return 0; }
+
+	// In a HID Class Descriptor wIndex cointains the interface number
+	if (setup.wIndex != pluggedInterface) { return 0; }
+
+	// Reset the protocol on reenumeration. Normally the host should not assume the state of the protocol
+	// due to the USB specs, but Windows and Linux just assumes its in report mode.
+	protocol = HID_REPORT_PROTOCOL;
+
+	// Transmit HID descriptor. See comments next to the descriptor parts
+	int ret = 0;
+	ret += USB_SendControl(TRANSFER_PGM, _hidReportDescriptorTouchscreen_1, sizeof(_hidReportDescriptorTouchscreen_1));
+	for (int i = 0; i < HID_MULTITOUCH_REPORTFINGERS; i++) {
+		ret += USB_SendControl(TRANSFER_PGM, _hidReportDescriptorTouchscreen_2, sizeof(_hidReportDescriptorTouchscreen_2));
+	}
+	ret += USB_SendControl(TRANSFER_PGM, _hidReportDescriptorTouchscreen_3, sizeof(_hidReportDescriptorTouchscreen_3));
+
+	return ret;
+}
+
+bool MultiTouch_::setup(USBSetup& setup)
+{
+	if (pluggedInterface != setup.wIndex) {
+		return false;
+	}
+
+	uint8_t request = setup.bRequest;
+	uint8_t requestType = setup.bmRequestType;
+
+	if (requestType == REQUEST_DEVICETOHOST_CLASS_INTERFACE)
+	{
+		if (request == HID_GET_REPORT) {
+			if(setup.wValueH == HID_REPORT_TYPE_FEATURE){
+				// The only feature is Contact Count Maximum
+				USB_SendControl(0, featureReport, featureLength);
+				return true;
+			}
+			return true;
+		}
+		if (request == HID_GET_PROTOCOL) {
+			// TODO improve
+#ifdef __AVR__
+			UEDATX = protocol;
+#endif
+			return true;
+		}
+		if (request == HID_GET_IDLE) {
+			// TODO improve
+#ifdef __AVR__
+			UEDATX = idle;
+#endif
+			return true;
+		}
+	}
+
+	if (requestType == REQUEST_HOSTTODEVICE_CLASS_INTERFACE)
+	{
+		if (request == HID_SET_PROTOCOL) {
+			protocol = setup.wValueL;
+			return true;
+		}
+		if (request == HID_SET_IDLE) {
+			idle = setup.wValueL;
+			return true;
+		}
+		if (request == HID_SET_REPORT)
+		{
+			// Check if data has the correct length afterwards
+			int length = setup.wLength;
+
+			// Feature (set feature report)
+			if(setup.wValueH == HID_REPORT_TYPE_FEATURE){
+				// No need to check for negative featureLength values,
+				// except the host tries to send more then 32k bytes.
+				// We dont have that much ram anyways.
+				if (length == featureLength) {
+					USB_RecvControl(featureReport, featureLength);
+					return true;
+				}
+				// TODO fake clear data?
+			}
+
+			// Output (not applicable)
+			else if(setup.wValueH == HID_REPORT_TYPE_OUTPUT){
+				return true;
+			}
+
+			// Input (set HID report)
+			else if(setup.wValueH == HID_REPORT_TYPE_INPUT)
+			{
+				if(length == sizeof(_report)){
+					USB_RecvControl(&_report, length);
```

Comment:
Will the PC send back a report? I do not think this will ever happen, does it?

Reply:
I don't think so. It's just that BootKeyboard has it like this. Actually I don't think any of the Host2Device HID_SET_REPORT control requests make sense (except the input messages for BookKeyboard leds). I can remove them but I'm just not 100% sure that they are unnecessary.

Reply:
It is a special case for bootkeyboard. You can remove it. Just test and if it works, keep it like that.

Reply:
Removed.